rbf-chunker/src/main.rs

```diff
@@ -0,0 +1,226 @@
+use std::fs::File;
+use std::io::prelude::*;
+use std::io::{Error, ErrorKind};
+use std::path::{Path, PathBuf};
+
+#[derive(Debug, Copy, Clone)]
+enum SectionType {
+    Unknown,
+    Header,
+    Footer,
+    Packet { crc: [u8; 2] },
+    PacketCRC,
+}
+
+#[derive(Debug, Copy, Clone)]
+struct SectionInfo {
+    offset: usize,
+    size: usize,
+    stype: SectionType,
+}
+
+impl SectionInfo {
+    fn new_header(size: usize) -> Self {
+        SectionInfo {
+            offset: 0,
+            size,
+            stype: SectionType::Header,
+        }
+    }
+
+    fn new_unknown(offset: usize, size: usize) -> Self {
+        SectionInfo {
+            offset,
+            size,
+            stype: SectionType::Unknown,
+        }
+    }
+
+    fn new_footer(offset: usize, size: usize) -> Self {
+        SectionInfo {
+            offset,
+            size,
+            stype: SectionType::Footer,
+        }
+    }
+
+    fn new_packet(offset: usize, size: usize, crc: [u8; 2]) -> Self {
+        SectionInfo {
+            offset,
+            size,
+            stype: SectionType::Packet { crc },
+        }
+    }
+
+    fn new_packet_crc(offset: usize) -> Self {
+        SectionInfo {
+            offset,
+            size: 2,
+            stype: SectionType::PacketCRC,
+        }
+    }
+}
+
+const PACKET_MIN_SIZE: usize = 916;
+const PACKET_MAX_SIZE: usize = 916;
+const DISCARD_MAX_SIZE: usize = PACKET_MAX_SIZE;
+const HEADER_SIZE: usize = 0x84;
+const FOOTER_SIZE: usize = 0x268;
+const EXPECTED_UNKNOWN_SIZE: usize = 0x378; // Used only for optimization.
+
+fn partition_rbf(
+    id: PathBuf,
+    rbf: &[u8],
+) -> Result<Vec<SectionInfo>, (Vec<SectionInfo>, &'static str)> {
+    let mut section_infos = Vec::with_capacity(
+        (rbf.len() - EXPECTED_UNKNOWN_SIZE - HEADER_SIZE - FOOTER_SIZE) / PACKET_MIN_SIZE * 2 + 3,
+    );
+    section_infos.push(SectionInfo::new_header(HEADER_SIZE));
+
+    let mut discard_start = HEADER_SIZE;
+    let mut section_start = HEADER_SIZE;
+    let mut section_end = HEADER_SIZE;
+    let mut running_crc = crc16::State::<crc16::MODBUS>::new();
+
+    while section_end != rbf.len() {
+        if section_start == section_end {
+            section_end += PACKET_MIN_SIZE;
+            if section_end > rbf.len() {
+                // footer found
+                break;
+            }
+            running_crc.update(&rbf[section_start..section_end - 2]);
+        } else if section_end - section_start < PACKET_MAX_SIZE {
+            section_end += 1;
+            running_crc.update(&[rbf[section_end - 1 - 2]]);
+        } else if section_start - discard_start >= DISCARD_MAX_SIZE {
+            section_infos.push(SectionInfo::new_unknown(
+                discard_start,
+                section_start - discard_start,
+            ));
+            return Err((section_infos, "No packet found. Refusing to continue."));
+        } else {
+            running_crc = crc16::State::<crc16::MODBUS>::new();
```

Comment:
I will note that the new discard logic is wholly contained in these 4 lines right here (a lot simpler then the mess in the v1/v2). I haven't bothered disabling it for non-first packets as I imagine that would cost more effort then it is worth.

Reply:
Sounds good to me!